### PR TITLE
[Bug fix] Fix Llama-3.1-8B eval accuracy under vLLM async scheduling

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
### Summary
Under --async-scheduling, Llama-3.1-8B produced incorrect decode output at reset / batch-condense steps, regressing the eval/accuracy tests. The main fix restores correct generation by keeping the authoritative device token across decode resets. Two supporting fixes restore seeded-sampling determinism on multi-chip Blackhole (which also unblocks the seed=0 chat-completions test).

### Main fix - keep the device "async-ahead" token across decode reset (generator.py)
Under async scheduling, the host-supplied last token lags device sampling at reset / prefill→decode mode-switch steps. Staging that stale host token (and its lagging position) re-processes a token at the wrong position - overwriting KV / regenerating positions - which produces duplicate or flipped tokens under concurrency and degrades eval accuracy.

This keeps the authoritative device token (and its position) for continuing slots, taking host tokens only for slots freshly prefilled since the last decode submit (_slots_prefilled_since_decode). slot_remap carries global slot indices (the plugin offsets each DP rank's local [0,B) map by rank*B), so it's rebased to per-rank local before gathering dev_toks/dev_pos, otherwise rank ≥1 indexes out of bounds. Position pairing is gated by TT_FIX_RESET_POS (default on).

#### Restore seed registration + RNG-counter↔position alignment (generator.py)
Main #45166 ("Clean up sampling separation in generators") dropped reset_seed_from_slots_if_needed + align_seed_counters_to_positions from sample_decode_on_device. Without them the per-request seed never reaches the device (the seed manager stays unseeded) and the RNG counter isn't tied to the absolute decode position, so seeded sampling is non-deterministic under async scheduling. Restores both calls before get_new_values, matching the known-good path. (This is what was failing the seed=0 test; the SeedManager methods themselves were untouched - they just stopped being called.)

#### Disable batched prefill on multi-chip Blackhole (model_config.py, #4004)
Prefill logits are batch-variant on multi-chip Blackhole (float-reduction order depends on batched-token count), so same-seed requests in different-sized prefill batches diverge. Disabled for Llama-3.1-8B on P300/P150x4/P150x8 (alongside main's existing Qwen3-32B/P150x4 case).

### Potential impact on other models
The two generator.py changes live in the shared tt_transformers.Generator, so Qwen3-32B, Mistral, Phi, gemma3, gemma4 and gpt_oss inherit them too (positive for seeded sampling, no-op when unseeded). Not affected: llama3_70b_galaxy (already has the alignment - it's the reference) and deepseek_v3 (own generator, lacks the alignment independently — flagged for its owners).

### Potentially important information for reviewers
The seed-registration change restores behavior #45166 intentionally removed across all shared-generator models, so might need to inform PR's author (Tomasz Cheda) so it isn't dropped again.

### CI runs
INFO: Gemma 4 fix is on current main, this branch doesn't contain it, so because of this there are Gemma 4 fails on tests, but this fix doesn't even target this model

CI evals + seeding tests: https://github.com/tenstorrent/tt-shield/actions/runs/27765421449/job/82176698553
vLLM nightly: https://github.com/tenstorrent/tt-metal/actions/runs/27815973961
Tier 1 + 2 E2E tests: https://github.com/tenstorrent/tt-metal/actions/runs/27818140967

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-ifabijanic/async-decode-reset)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-ifabijanic/async-decode-reset)

### CODE REVIEW

File: models/tt_transformers/tt/model_config.py [hunk 1]
```diff
@@ -597,15 +597,17 @@ def __init__(
...
         # sampling under concurrency then diverges (tt-inference-server#4004,
         # test_non_uniform_seeding). Forcing per-user batch-1 prefill removes the
-        # variance. Workaround until the prefill kernels are batch-invariant;
-        # mirrors the existing Llama-3.1-8B P300/P150x8 handling.
-        self.disable_batched_prefill = self.base_model_name == "Qwen3-32B" and self.device_name == "P150x4"
+        # variance. Workaround until the prefill kernels are batch-invariant.
+        # Disabled for Qwen3-32B (P150x4) and Llama-3.1-8B (P300/P150x4/P150x8).
+        self.disable_batched_prefill = (
+            self.base_model_name == "Qwen3-32B" and self.device_name == "P150x4"
+        ) or (self.base_model_name == "Llama-3.1-8B" and self.device_name in ("P300", "P150x4", "P150x8"))
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/tt_transformers/tt/model_config.py]: Update this code per suggestion.

**Review by mtairum — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -1224,6 +1233,77 @@ def decode_forward(
...
+        ):
+            new_tokens = []
+            new_start_pos = []
+            # Correctness fix (on by default; set TT_FIX_RESET_POS=0 to disable for
+            # A/B). When we take the device's async-ahead token for a continuing
+            # slot, the token sits at position dev_pos; staging the lagging host
+            # position (=dev_pos-1) re-processes that token at the wrong position
+            # (overwriting KV / regenerating a position -> duplicate/flipped tokens
+            # under concurrency). Pair the device token with the device position.
+            _fix_reset_pos = os.environ.get("TT_FIX_RESET_POS", "1") != "0"
```
Comment (djordje-tt) [inline models/tt_transformers/tt/generator.py]: Is this needed? 
What shall we set this flag for?

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -1224,6 +1233,77 @@ def decode_forward(
         tokens = torch.chunk(tokens, self.data_parallel, 0)
         start_pos = torch.chunk(start_pos, self.data_parallel, 0)
         page_table = torch.chunk(page_table, self.data_parallel, 0) if page_table is not None else None
+
+        # vLLM under async scheduling supplies a one-step-stale last token at
```
Comment (djordje-tt) [inline models/tt_transformers/tt/generator.py:1237]: So this happens with vllm, we have 1-step-stale last token. What happens when it's called through tt-metal? Do we still call this? Or is it properly passing then .

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (ifabijanic-tt) [reply models/tt_transformers/tt/generator.py]: This was a leftover from A/B testing during bug fixing, and you are right this is definitely not needed. Thanks for the info.

File: models/tt_transformers/tt/generator.py (hunk 3 above)
Comment (ifabijanic-tt) [reply models/tt_transformers/tt/generator.py:1237]: As much as I can see that block is vLLM-only in practice. On the direct tt-metal path it should never execute so the token is passed correctly there and this path should be OK.

The block is guarded by four conditions that must all hold:

_on_device_sampling and (reset_batch or mode_switched) and enable_trace and self.trace_inputs_decode[on_device_sampling]_

The 4th is the gate. self.trace_inputs_decode is a defaultdict(lambda: None) that only gets populated during the first traced decode capture which happens after this block, inside _decode_forward_trace_text.
The tt-metal demo calls with _reset_batch=(iteration == 0):_

Iteration 0 should have _reset_batch=True and mode_switched=True_, but this is the first decode so _trace_inputs_decode_ is still None so the block is skipped. The trace gets captured during this call.

All iteration after will have reset_batch=False and already in DECODE mode (mode_switched=False) → (reset_batch or mode_switched) is False → block skipped.
So the demo never has a step where a reset/mode-switch coincides with an already-captured decode trace (ar at least I cannot see it or trigger it).
Only vLLM should hit that: it re-submits batches mid-stream (reset_batch=True for batch condense and slot remaps) after decode has been running, and under --async-scheduling its host token lags the device by one step.
